src/equihash.rs

```diff
@@ -0,0 +1,321 @@
+use std::mem;
+use std::ptr;
+
+use byteorder::{LittleEndian, WriteBytesExt};
+use libsodium_sys::crypto_generichash_blake2b_init_salt_personal;
+use libsodium_sys::crypto_generichash_blake2b_state;
+use libsodium_sys::crypto_generichash_blake2b_update;
+use libsodium_sys::crypto_generichash_blake2b_final;
+use libsodium_sys::crypto_generichash_statebytes;
+
+struct Params {
+    n: u32,
+    k: u32,
+}
+
+struct Node {
+    hash: Vec<u8>,
+    indices: Vec<u32>,
+}
+
+impl Params {
+    fn indices_per_hash_output(&self) -> u32 {
+        512 / self.n
+    }
+    fn hash_output(&self) -> usize {
+        (self.indices_per_hash_output() * self.n / 8) as usize
+    }
+    fn collision_bit_length(&self) -> usize {
+        (self.n / (self.k + 1)) as usize
+    }
+    fn collision_byte_length(&self) -> usize {
+        (self.collision_bit_length() + 7) / 8
+    }
+    fn hash_length(&self) -> usize {
+        ((self.k as usize) + 1) * self.collision_byte_length()
+    }
+}
+
+impl Node {
+    fn new<'a>(a: &'a mut Node, b: &'a mut Node, trim: usize) -> Self {
+        let mut hash = vec![0u8; a.hash.len() - trim];
+        let mut indices = Vec::new();
+        for i in trim..a.hash.len() {
+            hash[i - trim] = a.hash[i] ^ b.hash[i];
+        }
+        if a.indices_before(b) {
+            indices.append(a.indices.as_mut());
+            indices.append(b.indices.as_mut());
+        } else {
+            indices.append(b.indices.as_mut());
+            indices.append(a.indices.as_mut());
+        }
+        Node {
+            hash: hash,
+            indices: indices,
+        }
+    }
+
+    fn indices_before(&self, other: &Node) -> bool {
+        // Indices are serialized in big-endian so that integer
+        // comparison is equivalent to array comparison
+        self.indices[0] < other.indices[0]
+    }
+
+    fn is_zero(&self, len: usize) -> bool {
+        for i in 0..len {
+            if self.hash[i] != 0 {
+                return false;
+            }
+        }
+        return true;
+    }
+}
+
+impl Clone for Node {
+    fn clone(&self) -> Self {
+        Node {
+            hash: self.hash.clone(),
+            indices: self.indices.clone(),
+        }
+    }
+}
+
+fn create_state() -> *mut crypto_generichash_blake2b_state {
+    let mut st = vec![0u8; (unsafe { crypto_generichash_statebytes() })];
+    unsafe { mem::transmute::<*mut u8, *mut crypto_generichash_blake2b_state>(st.as_mut_ptr()) }
```

Comment:
You need to leak `st` (std::mem::forget) or the `Vec` will deallocate and this function will always return a dangling pointer.

If I were you, I would encapsulate all of this so that we can be sure there are no memory safety problems. Create a structure containing `*mut crypto_generichash_blake2b_state`. It'll be a transmuted pointer from a `Box::into_raw`, which you can get with `Vec::into_boxed_slice` if you want to keep `crypto_generichash_statebytes` dynamic. You can `Box::from_raw` to get the box back in your `Drop` impl.

Also, use `std::mem::align_of` to test that `u8` and `crypto_generichash_blake2b_state` have the same alignment.

Reply:
Replaced with `blake2b` crate.